### PR TITLE
refactor(#49): make InitializationError more specific

### PR DESCRIPTION
**Pull request by 0xsecaas**
This commit refactors the `InitializationError` enum to be more specific about the error conditions. Instead of a generic `BadParameters(String)` error, we now have more specific errors like `DuplicateParticipants`, `NotEnoughParticipants`, `ThresholdTooSmall`, `ThresholdTooLarge`, and `InvalidParticipantIndex`.

This makes the code easier to debug and review, as the error type now carries more information about what went wrong.

This commit also adds a unit test to ensure that the correct error is returned when duplicate participants are provided.

### CODE REVIEW

File: src/protocol/errors.rs [hunk 1]
```diff
@@ -71,8 +71,20 @@ impl From<Box<dyn error::Error + Send + Sync>> for ProtocolError {
...
-    #[error("bad parameters: {0}")]
-    BadParameters(String),
+    #[error("bad parameter: {0}")]
+    BadParameter(String),
+    #[error("participant list cannot contain duplicates")]
+    DuplicateParticipants,
+    #[error(
+        "threshold {threshold} cannot be larger than the number of participants {participants}"
+    )]
+    NotEnoughParticipants { threshold: u32, participants: u32 },
```
Comment (gilcu3) [inline src/protocol/errors.rs:83]: I see that you tried to generalize the error of not having enough participants, and not having enough participants for a given threshold. I would rather keep the separate
```suggestion
    #[error( "participant count cannot be < 2, found: {participants}" )]
    NotEnoughParticipants { participants: u32 },
    #[error(
        "not enough intersecting old/new participants ({participants}) to reconstruct private key for resharing with threshold bigger than old threshold ({threshold})"
    )]
    NotEnoughParticipantsForThreshold { threshold: u32, participants: u32 },
```

File: src/generic_dkg.rs [hunk 2]
```diff
@@ -597,52 +597,41 @@ pub(crate) fn reshare_assertions<C: Ciphersuite>(
...
+    let old_participants =
+        ParticipantList::new(old_participants).ok_or(InitializationError::DuplicateParticipants)?;
 
     if old_participants.intersection(&participants).len() < old_threshold {
-        return Err(InitializationError::BadParameters(
-            format!("not enough intersecting old/new participants {:?} to reconstruct private key for resharing with threshold bigger than old threshold {}",
-            old_participants.intersection(&participants),
-            old_threshold).to_string(),
-        ));
+        return Err(InitializationError::NotEnoughParticipants {
```
Comment (gilcu3) [inline src/generic_dkg.rs:625]:
```suggestion
        return Err(InitializationError::NotEnoughParticipantsForThreshold {
```

File: src/generic_dkg.rs [hunk 3]
```diff
@@ -597,52 +597,41 @@ pub(crate) fn reshare_assertions<C: Ciphersuite>(
     old_participants: &[Participant],
 ) -> Result<(ParticipantList, ParticipantList), InitializationError> {
     if participants.len() < 2 {
-        return Err(InitializationError::BadParameters(format!(
-            "participant count cannot be < 2, found: {}",
-            participants.len()
-        )));
+        return Err(InitializationError::NotEnoughParticipants {
+            threshold: 2,
```
Comment (gilcu3) [inline src/generic_dkg.rs:601]:
```suggestion
```

File: src/eddsa/sign.rs [hunk 4]
```diff
@@ -212,26 +212,24 @@ pub fn sign(
     message: Vec<u8>,
 ) -> Result<impl Protocol<Output = SignatureOutput>, InitializationError> {
     if participants.len() < 2 {
-        return Err(InitializationError::BadParameters(format!(
-            "participant count cannot be < 2, found: {}",
-            participants.len()
-        )));
+        return Err(InitializationError::NotEnoughParticipants {
+            threshold: 2,
```
Comment (gilcu3) [inline src/eddsa/sign.rs:216]:
```suggestion
```

**Review by gilcu3 — CHANGES_REQUESTED** [review]
Thank you very much!
Mostly minor changes requested, will check again once CI works and it is rebased on main.

File: src/confidential_key_derivation/protocol.rs [hunk 5]
```diff
@@ -271,4 +268,33 @@ mod test {
...
+            me,
+            private_share,
+            app_id,
+            app_pk,
+        );
+        match result {
+            Ok(_) => panic!("Expected an error, but got Ok"),
+            Err(err) => assert_eq!(err, InitializationError::DuplicateParticipants),
+        }
+    }
```
Comment (SimonRastikian) [inline src/confidential_key_derivation/protocol.rs:299]: Nit: Replace this by -> Result<(); dyn Box> in the function signature then call ckd and add ? operator at the end.

**Comment by SimonRastikian** [comment]
All good!
